auth_session_timeout/models/res_users.py

```diff
@@ -0,0 +1,109 @@
+        """
+        params = self.env['ir.config_parameter']
+        return params._auth_timeout_get_parameter_ignored_urls()
+
+    @api.model_cr_context
+    def _auth_timeout_deadline_calculate(self):
+        """Pluggable method for calculating timeout deadline
+        Defaults to current time minus delay using delay stored as config
+        param.
+        """
+        params = self.env['ir.config_parameter']
+        delay = params._auth_timeout_get_parameter_delay()
+        if delay <= 0:
+            return False
+        return time() - delay
+
+    @api.model_cr_context
+    def _auth_timeout_session_terminate(self, session):
+        """Pluggable method for terminating a timed-out session
+
+        This is a late stage where a session timeout can be aborted.
+        Useful if you want to do some heavy checking, as it won't be
+        called unless the session inactivity deadline has been reached.
+
+        Return:
+            True: session terminated
+            False: session timeout cancelled
+        """
+        if session.db and session.uid:
+            session.logout(keep_db=True)
+        return True
+
+    @api.model_cr_context
+    def _auth_timeout_check(self):
+        """Perform session timeout validation and expire if needed."""
+
+        if not http.request:
+            return
+
+        session = http.request.session
+
+        # Calculate deadline
+        deadline = self._auth_timeout_deadline_calculate()
+
+        # Check if past deadline
+        expired = False
+        if deadline is not False:
+            path = http.root.session_store.get_session_filename(session.sid)
+            try:
+
+                expired = getmtime(path) < deadline
+            except OSError as e:
+                _logger.exception(
+                    'Exception reading session file modified time.',
+                )
+                # Force expire the session. Will be resolved with new session.
+                expired = True
+
+        # Try to terminate the session
+        terminated = False
+        if expired:
+            terminated = self._auth_timeout_session_terminate(session)
+
+        # If session terminated, all done
+        if terminated:
+            return
+
+        # Else, conditionally update session modified and access times
+        ignored_urls = self._auth_timeout_get_ignored_urls()
+
+        if http.request.httprequest.path not in ignored_urls:
+            if 'path' not in locals():
+                path = http.root.session_store.get_session_filename(
+                    session.sid,
+                )
+            try:
+                utime(path, None)
+            except OSError as e:
+                _logger.exception(
+                    'Exception updating session file access/modified times.',
+                )
+
+    @classmethod
+    def check(cls, *args, **kwargs):
+        res = super(ResUsers, cls).check(*args, **kwargs)
+        http.request.env.user._auth_timeout_check()
```

Comment:
@nadiaafa If you do a RPC request then this fails with this error (see LRMIG-679 intern):

```
   File "/opt/odoo/odoo/service/wsgi_server.py", line 124, in wsgi_xmlrpc
     result = odoo.http.dispatch_rpc(service, method, params)
   File "/opt/odoo/odoo/http.py", line 115, in dispatch_rpc
     result = dispatch(method, params)
   File "/opt/odoo/odoo/service/model.py", line 35, in dispatch
     security.check(db,uid,passwd)
   File "/opt/odoo/odoo/service/security.py", line 13, in check
     return res_users.check(db, uid, passwd)
   File "/opt/odoo/addons_extern/auth_session_timeout/models/res_users.py", line 108, in check
     http.request.env.user._auth_timeout_check()
   File "/usr/local/lib/python3.5/dist-packages/werkzeug/local.py", line 343, in __getattr__
    return getattr(self._get_current_object(), name)
   File "/usr/local/lib/python3.5/dist-packages/werkzeug/local.py", line 302, in _get_current_object
     return self.__local()
   File "/usr/local/lib/python3.5/dist-packages/werkzeug/local.py", line 135, in _lookup
     raise RuntimeError('object unbound')
 RuntimeError: object unbound
```

I fixed this in our project with replacing:

```python
    @classmethod
    def check(cls, *args, **kwargs):
        res = super(ResUsers, cls).check(*args, **kwargs)
        if http and http.request and http.request.env and \
                http.request.env.user:
            http.request.env.user._auth_timeout_check()
        return res
```